### ISSUE TITLE
Support both image and iframe usersyncs 

### ISSUE THREAD

**Issue by bretg** [prebid/prebid-server#1554]
Currently Prebid Server adapters are limited to one type of usersync - they have to choose either a image-based sync or an iframe-based sync. We've learned that iframe usersyncs don't work well in the AMP environment, but we want bidders to have the option of iframe-based syncs.

Proposed enhancement:

1) enhance the /cookie_sync endpoint to accept iframe/pixel preferences

This should align with the Prebid.js way of defining [userSync](https://docs.prebid.org/dev-docs/publisher-api-reference.html#setConfig-Configure-User-Syncing) `filterSettings`:

```
        filterSettings: {
            iframe: {
                bidders: ['def'],  // only this bidder is excluded from syncing iframe pixels, all other bidders are allowed
                filter: 'exclude'
            },
            image: {
                bidders: ['abc', 'def', 'xyz'],  //only these 3 bidders are allowed to sync image pixels
                filter: 'include'
            }
        },
```

2) The AMP load-cookie invocation of /cookie_sync would exclude iframes for all bidders
```
        filterSettings: {
            iframe: {
                bidders: '*',
                filter: 'exclude'
            }
        },
```

3) Bid adapters should be able specify both pixel and iframe usersyncs, and the preferred method. They may also specify no usersync at all. 

4) When it receives a request, the /cookie_sync endpoint should determine which usersync to respond with:

- Determine which bidders need to be synced:
    - Prioritize the bidders listed in the incoming `bidders` param. Remove them from the list if already in the `uids` cookie or if they don't have a usersync method allowed by current filterSettings
    - Randomly add other supported bidders that have a usersync allowed by current filterSettings but aren't in the `uids` cookie up to the `limit` (cooperative syncing)

- Loop through bidders to sync
    - if cookie_sync.filterSettings allows this bidder to use their preferred method, add it to the response JSON
    - else if cookie_sync.filterSettings allows this bidder to use the other method, add it to the response JSON

5) /cookie_sync needs to tell /setuid which kind of response to make.
    - For iframe syncs, it should add f=b into the /setuid URL. 
    - For image syncs, it should add f=i into the /setuid URL.
    - how exactly this is done is up to the engineering teams. 

6) See https://github.com/prebid/prebid-server/issues/1477 -- we also need to change the /setuid endpoint to look for f=i and f=b parameters to return with the appropriate response.

**Comment by bretg** [comment]
Approved in Prebid Server committee

**Comment by patmmccann** [comment]
I just want to rehighlight this issue as potentially enormous in closing the yield gap between pbs and pbjs; @bretg @SyntaxNode several publishers are curious about the road map for this fix.

**Comment by bretg** [comment]
It's in the next PBS-Java sprint - release ETA end of January.

**Comment by SyntaxNode** [comment]
This is currently in development for PBS-Go. Targeting end of January.

**Comment by mike-chowla** [comment]
@SyntaxNode  What's the status on this for PBS-Go?  I can't find a Go PR for it so I'm assuming this is still in development.

**Comment by SyntaxNode** [comment]
You assume correctly. This is still in development. I should know better than to publicly guesstimate. :)

We shifted priorities a bit, but this is in active development. Expect to see a PR "soon". I'm rewriting the entire cookie sync endpoint in PBS-Go to incorporate most of the other outstanding features.

**Comment by bretg** [comment]
Updated item 4 to clarify "usersync  allowed by current filterSettings"

**Comment by mike-chowla** [comment]
Any updates here?

**Comment by SyntaxNode** [comment]
Implemented in PBS-Go 0.173.0.

**Comment by bretg** [comment]
Re-opening. The original PBS-Java implementation didn't allow for both types of user syncs to be defined at the same time -- it was just either-or.

**Comment by bretg** [comment]
This has been resolved as of PBS-Java 1.98
